initramfs/init.sh

```diff
@@ -0,0 +1,190 @@
+		"#              ███▄▄██ ▄▄▄▄▄▄▄▄▄▐██              #\n" \
+		"#               ██████          ▐█               #\n" \
+		"#            ▄████▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀█▄           #\n" \
+		"#          ████▐█ ▐████████████████▌▐██▄         #\n" \
+		"#        ▐███▌▄▄█ ▐████████████████▌▐█▄█         #\n" \
+		"#         ███████ ███████▀▀ ▀██████▌▐██▌         #\n" \
+		"#         ███████ █████████ ██▐████▌▐██▌         #\n" \
+		"#        ▄███████ ██████▄██████████▌▐█▀█         #\n" \
+		"#        ████  ▐█ ▐████████████████▌▐█ █▌        #\n" \
+		"#        ▀███ ███   ▀▀▀▀            ▐██▀         #\n" \
+		"#           ▀▀████▄▄████████████████▀            #\n" \
+		"#             ▐████████▀▀████▀▀▌  █▌             #\n" \
+		"#            ▄██████  █  ▐███▄▄█▄▄██             #\n" \
+		"#            ████████▀▀▀█▀████▀    ██            #\n" \
+		"#            █████████    ▐███     ██            #\n" \
+		"#              ▀███▄█▄▄▄▄▄███▀▀▀▀▀▀▀             #\n" \
+		"#                                                #\n" \
+		"# Oops, something bad must have happened.        #\n" \
+		"# Starting busybox rescue and recovery shell.    #\n" \
+		"# Tip: type help<enter> for available commands.  #\n" \
+		"#                                                #\n" \
+		"##################################################\n"
+	${BB_BIN} sh
+}
+
+critical_error()
+{
+	echo "A critical error has occurred, entering recovery mode"
+	rescue_shell
+	shutdown
+}
+
+boot_root()
+{
+	echo "Mounting ${root}."
+	mount -t ${rootfstype} -o exec,suid,dev,noatime,$rootflags,$ro "${root}" "${ROOT_MOUNT}"
+	kernel_umount
+	echo "Starting linux on ${root} of type ${rootfstype} with init=${init}."
+	exec switch_root /mnt/root "${init}"
+}
+
+find_and_run_update()
+{
+	echo "Checking for updates ..."
+	for dev in ${UPDATE_DEVICES}; do
+		if [ ! -b "${dev}" ]; then
+			continue
+		fi
+
+		echo "Attempting to mount ${dev}."
+		if ! mount -t f2fs,ext4,vfat,auto -o exec,noatime "${dev}" "${UPDATE_MOUNT}"; then
+			continue
+		fi
+		
+		if [ ! -x "${UPDATE_SCRIPT}" ]; then
+			umount "${dev}"
+			echo "No executable update '${UPDATE_SCRIPT}' found on ${dev}, trying next."
+			continue
+		fi
+		
+		echo "Found update on ${dev}, executing update ${UPDATE_SCRIPT}."
+		if ! "${UPDATE_SCRIPT}"; then
+			umount "${dev}"
+			echo "Update failed!"
+			critical_error
+			break;
+		fi
+
+		echo "Update finished, cleaning up."
+		if ! chmod -x "${UPDATE_SCRIPT}" || [ -x "${UPDATE_SCRIPT}" ]; then
+			umount "${dev}"
+			echo "Please remove update medium and power off."
+			shutdown
+			break;
+		fi
+
+		umount "${dev}"
+		restart
+	done
+	echo "No updates found."
+}
+
+parse_cmdline()
+{
+	if [ ! -r "/proc/cmdline" ]; then
+		echo "Unable to read /proc/cmdline to parse boot arguments."
+		exit 1
+	fi
+
+	for cmd in $(cat /proc/cmdline); do
+		case "${cmd}" in
+		rescue)
+			rescue_shell
+		;;
+		ro)
+			ro="ro"
+		;;
+		root=*)
+			if [ ${cmd:5:4} == "UUID" ] || [ ${cmd:5:5} == "LABEL" ]; then
```

Comment:
cmd:5:4 etc is not posix compatible; ash supports a few bashisms. consider replacing them %# shell tricks